packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -0,0 +1,182 @@
+/* eslint-disable unicorn/prefer-add-event-listener */
+// cSpell:ignore njson
+
+import {
+  AsyncReturnType,
+  Cardano,
+  EraSummary,
+  NetworkInfoMethods,
+  NetworkInfoProvider,
+  StakeSummary,
+  SupplySummary,
+  WSMessage
+} from '@cardano-sdk/core';
+import { Logger } from 'ts-log';
+import { Observable, ReplaySubject, firstValueFrom } from 'rxjs';
+
+import { NJSON } from 'next-json';
+import WebSocket from 'isomorphic-ws';
+
+const NOT_CONNECTED_ID = 'not-connected';
+
+type WSStatus = 'connecting' | 'connected' | 'idle' | 'stop';
+
+export type WSHandler = (message: WSMessage) => void;
+
+export interface WsClientConfiguration {
+  /** The interval in seconds between two heartbeat messages. Default 55". */
+  heartbeatInterval?: number;
+
+  /** The interval in seconds after which a request must timeout. Default 60". */
+  requestTimeout?: number;
+
+  /** The WebSocket server URL. */
+  url: URL;
+}
+
+export interface WsClientDependencies {
+  /** The logger. */
+  logger: Logger;
+}
+
+const isEventError = (error: unknown): error is { error: Error } =>
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  typeof error === 'object' && !!error && (error as any).error instanceof Error;
+
+export class CardanoWsClient {
+  /** The client id, assigned by the server. */
+  clientId = NOT_CONNECTED_ID;
+
+  /** The `Observable` form of `NetworkInfoProvider`. */
+  networkInfo: {
+    [m in NetworkInfoMethods]: Observable<AsyncReturnType<NetworkInfoProvider[m]>>;
+  };
```

Comment:
Yes, it would be nice to generalize the type. I spent about 1h, but I wasn't able to.
There is the problem of extracting the `$` from the key to access the methods of the `Provider` and also it would be nice to generalize the `Observable`: we have two types (one public one private) almost identical.

Reply:
Let's capture this in a ticket when planning follow-up work

Reply:
LW-10837